drivers/sensor/tmp112/tmp112.c

```diff
@@ -194,6 +194,41 @@ static int tmp112_channel_get(const struct device *dev,
 	return 0;
 }
 
+#ifdef CONFIG_PM_DEVICE
+static int tmp112_pm_control(const struct device *dev, uint32_t command,
+			     uint32_t *state, pm_device_cb cb, void *arg)
+{
+	int ret = 0;
+	struct tmp112_data *device_data = get_data(dev);
+
+	switch (command) {
+	case PM_DEVICE_STATE_SET:
+		if (*state == PM_DEVICE_STATE_ACTIVE
+		    && device_data->power_state
+		    == PM_DEVICE_STATE_SUSPEND) {
+			tmp112_update_config(dev, 0, 0);
+			device_data->power_state = PM_DEVICE_STATE_ACTIVE;
+		} else if (*state == PM_DEVICE_STATE_SUSPEND) {
+			device_data->power_state = PM_DEVICE_STATE_SUSPEND;
```

Comment:
as far as I can see, only `PM_DEVICE_STATE_ACTIVE` is handled, if system requests `PM_DEVICE_STATE_SUSPEND` the driver does nothing but changing its state. How does device save power?

Reply:
On it's own it does not. But if the device can be powered on/off by external electronics then the state can be restored.

Reply:
A driver that implements PM needs to support at the very least active/suspend states. If the device can be powered on/off externally, then the device driver needs to know the actual device state, or, have control of such external electronics to change the device power states.

Reply:
> then the device driver needs to know the actual device state

Reply:
The problem I see here is that this is not actually implementing any device PM. If system is suspended, the device will get called with `PM_DEVICE_STATE_SET` and `PM_DEVICE_STATE_SUSPEND`. In this case, the device will do nothing other than setting it's state to `PM_DEVICE_STATE_SUSPEND`. As far as I understand, this is equivalent to have no device PM.

Reply:
The driver send the last known configuration when the pm state is set from `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_ACTIVE`. I record the current state in order to not send the configuration twice if the `PM_DEVICE_STATE_SET` is done twice in a row. We could save the configuration upon `PM_DEVICE_STATE_SUSPEND` but it's not required since `tmp112_attr_set` already does it

Reply:
I'm sorry, but I still don't understand how the device enters a _low power state_ if the system is suspended. `tmp112_attr_set` has nothing to do with device PM. Can you please detail the workflow?

Reply:
So let's say we have a board with several power domain. In one of those power domain, controlled by component `PWR`, we can find among other devices a `TMP112` device.

The `PWR` device driver (or the application) do know which components via (struct devices *) are under the power domain it controls. When the power domain is not in use anymore and need to be brought down, `TMP112` `pm_device_state_set(dev, PM_DEVICE_STATE_SUSPEND)` is called and then the power is turned-off.

When the power domain is required, i.e the application need access to one of the component. The power is turned on by `PWR` and  `pm_device_state_set(dev, PM_DEVICE_STATE_ACTIVE)` is called

Reply:
I'd expect to see something related to this sensor feature in the PM section:

![image](https://user-images.githubusercontent.com/25011557/122373232-9bf16d00-cf61-11eb-946e-7bea406a6e54.png)

What you are describing is really an application specific behavior as far as I understand.